### PR TITLE
Add an entity option to use fluent setters

### PR DESCRIPTION
**Pull request by cbornet**
Because fluent is much nicer !

### CODE REVIEW

**Comment by jdubois** [comment]
The entity one of the core JHipster object, can we discuss this before? I'm not familiar with having a "fluent" setter on a domain object - I see the point, of course, but I've never seen those used on JPA entities, and wonder if there are any drawbacks or issues. Besides, I'm not sure everyone will want those (as I've never seen those used for domain objects), so that would mean dead code for a lot of people. Of course, if there are only advantages, we could also evangelize people to use those more!

**Comment by deepu105** [comment]
Either we should use normal setters or fluent setters, having both might be lot of dead code. I don't see any downside in having fluent setters that follow the same naming convention like below
```
class Foo {

  private String bar;

  public String getBar() {
    return bar;
  }

  public Foo setBar(String bar) {
    this.bar = bar;
    return this;
  }
}
```

**Comment by cbornet** [comment]
I think setBar can lead to issues with JavaBean reflection so bar(String bar) would be better.

**Comment by cbornet** [comment]
I thought hibernate would complain if it didn't have the void setBar() method for its reflection but it seems to be OK so maybe we can remove it.

**Comment by deepu105** [comment]
@cbornet what about mapstruct does it play well with this syntax

**Comment by deepu105** [comment]
btw what issue do you foresee with JavaBean reflection?

**Comment by cbornet** [comment]
I don't know, something I read somewhere...
Anyway, we can test with setBar() and if it works, it's just a matter of preference between setBar() or bar().

**Comment by cbornet** [comment]
As for mapstruct, I didn't test directly but I guess that it was tested OK by CI.

**Comment by jdubois** [comment]
you need `setBar` otherwise you are not following the JavaBean convention. Speaking of this, you should also return `void`, so I'm really not sure we should do this.

**Comment by cbornet** [comment]
That's why I duplicated the method. But on in other hand, I removed the ```void setBar``` method and both Jackson and Hibernate seem to be fine with it.

**Comment by cbornet** [comment]
To show the interest of fluent setters, compare
```java
Foo foo = new Foo();
foo.setMyString("foo");
foo.setMyString2("bar");
fooRepository.save(foo);
```
with
```java
fooRepository.save( new Foo().myString("foo").myString2("bar") );
```
This is a simple example but it gets more and more interesting when structures get complex. Consider
```java
Foo foo = new Foo();
foo.setMyString("test");
fooRepository.save(foo);

Bar bar = new Bar();
bar.setMystr("bar");
bar.setFoo(foo);
barRepository.save(bar);
```
vs
```java
barRepository.save( new Bar()
    .foo(fooRepository.save(new Foo()
        .myString("foo1")
        .myString2("foo2")))
    .mystr("bar"));
```
With the fluent version, it almost reads as YAML !

Note: the repository ```save``` can probably be removed if the entities have CascadeType.ALL on their relationships.
@jdubois Is there a reason why we don't set the relationships as CascadeType.ALL by default ?

**Comment by jdubois** [comment]
- OK the fluent interface looks better, then I've never seen anyone using it -> that's why it worries me, and I wonder if there is a technical issue somewhere
- For CascadeType.ALL this is because of our [Policy 1](https://jhipster.github.io/policies/). Anyway here you have a different issue: you need to `save()` because that's a new entity, so it's non-managed. Otherwise it would work automatically (at least without the fluent interface, it works!!)

**Comment by cbornet** [comment]
I thought CascadeType.ALL (or persist) was used so that the entity is automatically added to the session. I need to check that again.

**Comment by cbornet** [comment]
If there is no JavaBean setter, Hibernate [uses properties directly even if they are private](http://docs.jboss.org/hibernate/orm/4.3/manual/en-US/html_single/) (unlike standard JPA) and Jackson will [both serialize and deserialize](http://www.baeldung.com/jackson-field-serializable-deserializable-or-not) (as long as there is a getter).
Now for the other frameworks that use the domain (mongo, ES, cassandra, liquibase ...) , I don't know the impact of not having a Javabean setter but it seems the CI tests where good.

My opinion would be to have both ```void setBar()``` and ```bar()``` generated to avoid issues with libraries that expect to have the JavaBean setter. JavaBean reflection will just ignore the ```bar()``` method.

**Comment by cbornet** [comment]
In java.beans.Introspector
```java
                       } else if (void.class.equals(resultType) && name.startsWith(SET_PREFIX)) {
                            // Simple setter
                            pd = new PropertyDescriptor(this.beanClass, name.substring(3), null, method);
                            if (throwsException(method, PropertyVetoException.class)) {
                                pd.setConstrained(true);
                            }
                        }
```
so the void result type is checked before adding the method to the PropertyDescriptor list.
Also a non-javabean method will just be ignored so it's safe to add them.

**Comment by jdubois** [comment]
- I'll let you have a look at `CascadeType.ALL` :-)
- Then we are going to have both `setBar()` and `bar()`, and once again maybe it can cause issues (JPA, MapStruct, Jackson... so many stuff use the domain object), and this is going to be dead code for a lot of people, so this is bad for them. At best this could be an option.

**Comment by cbornet** [comment]
I tested that
```
PropertyDescriptor descriptor = new PropertyDescriptor("bar", Foo.class);
```
fails with
```
java.beans.IntrospectionException: Method not found: setBar
```
If the void setBar method is not present

**Comment by cbornet** [comment]
It's not really dead code : it's code that people may not use (even if they should :smile: ) and it's in generated code so it's a bit like a library : you don't really need to go see the internal code often.

But if you prefer to do it as an option or module I understand. Please tell me which one in that case.

**Comment by cbornet** [comment]
This reminds me that I wanted to add adders and removers on relationships Set. So that could be part of the package.

**Comment by jdubois** [comment]
It's just that people keep telling me JHipster generates too much code, and that they don't understand what is being generated.
Of course, you could argue that if they don't understand the code, they shouldn't say there is too much code :-)

**Comment by cbornet** [comment]
So they want awesome features but no code. Like in magic ?

If you're ok, I'll do it as a silent option (no question asked) so that it can be configured in jhipster-uml/JDL

**Comment by deepu105** [comment]
--option might be good. So that people who knows and uses these stuff can
enable them when required

Thanks & regards,
Deepu

File: generators/entity/index.js [hunk 1]
```diff
@@ -207,6 +207,7 @@ module.exports = EntityGenerator.extend({
         this.changelogDate = this.fileData.changelogDate;
         this.dto = this.fileData.dto;
         this.service = this.fileData.service;
+        this.fluentMethods = this.fileData.fluentMethods;
```
Comment (deepu105) [inline generators/entity/index.js]: may be you can add `--fluentMethods` as an option to sub generator as well when creating entities by hand.
dont forget to PR the JDL project as well

**Comment by cbornet** [comment]
Yes, I'll add the command line option.

**Comment by cbornet** [comment]
Done. I'll add it to the JDL once this is merged.

**Comment by deepu105** [comment]
:+1:

**Comment by jdubois** [comment]
Then don't forget to document it on jhipster.github.io -> this is indeed a very nice option if you know what it does, and I'm sure a lot of people will like it
And can you also add "insight" statistics -> if a lot of people use it, we can then convert it to a new question

**Comment by cbornet** [comment]
Yup, will do. As for the insight, I have already added [this](https://github.com/cbornet/generator-jhipster/blob/3a760a472d4fa6e0184dd390005754cd09ae644d/generators/entity/index.js#L601). Is there something missing ?

Also what should be the semantic for the JDL ? ```fluent-methods A,B ``` ?

**Comment by deepu105** [comment]
@cbornet I think it should be `fluentMethods A,B` follow the syntax for `skipClient `

**Comment by cbornet** [comment]
ALse I just tried
```java
        Foo foo = new Foo()
            .myString("test")
            .addBar(new Bar().mystr("bar1"))
            .addBar(new Bar().mystr("bar2"));
        fooRepository.save(foo);
```
With a CascadeType.ALL on the OneToMany and the Bar entities where added to the DB.
If I don't put CascadeType.ALL, they are not.
So it does add the entities automatically without the need to call save.

**Comment by cbornet** [comment]
@deepu105 noted !

**Comment by jdubois** [comment]
For `CascadeType.ALL`-> yes, that's cascading the "many" entities, then that's because those are not already managed

I'd rather keep this with the default: I'd rather people learn how it works, than doing a surprise to people who already knows how it works

**Comment by cbornet** [comment]
No problem, it's not a big work to edit it by hand when you need it anyway.

Note : I should have wrote
```java
        fooRepository.save( new Foo()
            .myString("test")
            .addBar(new Bar().mystr("bar1"))
            .addBar(new Bar().mystr("bar2")));
```
:wink:

**Comment by ghahramani** [comment]
@jdubois Actually I converted all the jhipster setter to fluent style :D with no problems. I like this style it gets code much cleaner

**Comment by cbornet** [comment]
Remove "needs-discussion" label ?

**Comment by jdubois** [comment]
Yes let's do this, I agree with you!

**Comment by jdubois** [comment]
@cbornet maybe you could use those in some generated tests, so that people see how it should works? And if that's good, we could put this option as "true" by default... As we should push best practices to users, and that's definitely a good way to use the entities. My only concern is that we would need a lot of testing, to check if this works with all options (including MapStruct, Cassandra...)

**Comment by cbornet** [comment]
I have tested that javaBean reflection works correctly (both reading/writing) so if those libs use it they should be OK.
At the beginning I had done it in the tests but since this is now an option, I removed it. Now,  I could check the fluentMethods flag and generate the test accordingly.

**Comment by cbornet** [comment]
I have added the tests and now it's cross-tested by CI with Mongo and Cassandra.
I haven't added a cross-test with mapstruct but it is easy to do if needed.

**Comment by jdubois** [comment]
No need to rush -> if we do a release tomorrow, it's a bit early to merge this anyway.
But indeed I'd like to have this tested also with MapStruct. And we need to have MapStruct out of beta at some point :-)

**Comment by jdubois** [comment]
OK, this all looks good to me.

Now, 2 things:

- we need more testing -> but this can be done after merging this
- do we put this by default or not? I like the fact that we provide "best practices" and push them, then I don't know how people will react to this -> I'll do a Twitter survey and mention @cbornet

**Comment by jdubois** [comment]
And the vote has started at https://twitter.com/java_hipster/status/760504585748447232

**Comment by vladmihalcea** [comment]
Although the JPA specification forbids Java Bean conventions violation, Hibernate is not so strict. For instance, you can run [this test](https://github.com/vladmihalcea/high-performance-java-persistence/blob/master/core/src/test/java/com/vladmihalcea/book/hpjp/hibernate/mapping/FluentSettersTest.java) just fine. As long as you rely only on Hibernate, you should be fine. However, if you plan on supporting other JPA providers, the Fluent API might not work, so keep that in mind as well.

**Comment by jdubois** [comment]
Thanks a lot @vladmihalcea !!! I didn't know it was a JPA spec violation, and in that case that's quite a big issue to me. It would also go against our [Policy 1](http://jhipster.github.io/policies/).

**Comment by vladmihalcea** [comment]
It's the **2.2 Persistent Fields and Properties** section of the Java Persistence 2.1 spec that says:

> In this case, for every persistent property _property_ of type T of the entity, there is a getter method, _getProperty_, and setter method _setProperty_. For boolean properties, _isProperty_ may be used as an alternative name for the getter method.[2]
> 
> For single-valued persistent properties, these method signatures are:
> - T getProperty()
> - void setProperty(T t)

Now, the spec needs to be generic in order to allow interoperability. In reality, JPA implementations might relax some requirements to provide a better developer experience. 

Since you want to support any JPA provider, you might want to contact all major implementation and validate that the Fluent API is not breaking anything.

**Comment by jdubois** [comment]
Thanks again @vladmihalcea 
No we don't intend to support all JPA providers (just Hibernate, in fact), but we try to follow the "standard" usage of each technology (and so follow JPA, here).
I'll wait until the vote is over: at the moment we have 70% votes for the "fluent setters"!

**Comment by cbornet** [comment]
The PR doesn't remove the java bean methods and it must not as some frameworks may use java bean reflection. The fluent methods are added methods with different signature so they don't disturb the reflection.

**Comment by vladmihalcea** [comment]
But you cannot have two setter methods whose signature only differ in the return valuue type. How are you going to address that?

**Comment by cbornet** [comment]
They differ in name : ```void setBar()``` vs ```Foo bar()```.

**Comment by vladmihalcea** [comment]
Seems legit.

**Comment by jdubois** [comment]
So the vote is over, 116 votes, 72% for the "fluent setters".
Besides, we had a nice blog entry by @vladmihalcea about (see it [here](https://vladmihalcea.com/2016/08/09/fluent-api-entity-building-with-jpa-and-hibernate/)), and this was even tweeted by the official Java account (see it [here](https://twitter.com/java/status/763035625800949761)). So an overall very clear vote for this.

- let's do this
- let's not put this as an option: it adds more maintenance cost, and there is no big issue in doing this (except extra code for non-hipsters who want to use the classical setters)

**Comment by cbornet** [comment]
@jdubois great ! As you know I'm currently on the beach drinking mojitos without any keyboard nearby so can you take the PR from here ?

**Comment by jdubois** [comment]
Yes of course. I'm the one who broke it anyway!

**Comment by jdubois** [comment]
I've changed my mind after reading the code: I'm keeping the "fluent-method" option, for backward compatibility (and if people are unhappy with this), but I'm putting it at "true" by default.
It's a very simple switch, so no big maintenance cost in the end.

**Comment by jdubois** [comment]
I found a couple of small issues, I'll finish this tomorrow!

**Comment by jdubois** [comment]
I'm going to merge this, just a couple of issues I'd like to share before.

There was an "add*" method for adding an entity to a collection (for one-to-many and many-to-many relationships), so I added also a "remove*" method:

- Those "add*" and "remove*" methods forced me to add a new question in many-to-many relationships, when the entity is the owner -> you need to know the name of the relationship in the other entity. We were already asking this question for other use cases, and this is the "otherEntityRelationshipName" parameter.
- Those 2 methods also kind of force to have bi-directional relationships -> if you don't want the bi-directional relationship, there is a couple of more lines to delete

So for me this needs more testing and practice.

**Comment by cbornet** [comment]
I totally forgot about unidirectional relationships ! We should detect them and not set the remote field for these.

**Comment by jdubois** [comment]
Yes and I broke the build because of this... There must one edge case I missed.

This is quite annoying, I'm wondering if that's the right approach. I'm first correcting the build, but then we must have a look at this, and its consequences.

**Comment by jdubois** [comment]
In fact most of the errors where in our `.json` test files, which where not correct :-)

I've also added [a ticket to JHipster Core](https://github.com/jhipster/jhipster-core/issues/51) so this is included in the JDL!

**Comment by cbornet** [comment]
@jdubois I wanted to have a look at unidirectional relationships and I couldn't see any issue.
For unidir ManyToOne, the method is not generated.
And unidir OneToMany are not allowed in jhipster.
Am I missing something ?

**Comment by jdubois** [comment]
@cbornet I didn't try it after my last comment about it, and the code has changed since, so maybe there isn't any issue after all. If you tried and it works, that's perfect!

**Comment by cbornet** [comment]
Ok. I'll check when I come back.
Also I think there might still be an issue in ManyToMany : in the adder method we should add to the remote collection instead of calling its setter. Can someone have a look ?

**Comment by cbornet** [comment]
Oh ! You've already fixed that @jdubois ! 👍  😄

**Comment by jdubois** [comment]
I don't remember everything I've fixed :-)
But yes there was some serious work on this part! And I'm still not 100% sure it's perfect

**Comment by dhaalves-uel** [comment]
Check project Lombok https://projectlombok.org/features/index.html
This feature and more can be achieved with simple annotations

**Comment by cm325** [comment]
I was actually kind of suprised that jhipster wasn't already using lombok, seems like that's what all the cool kids are using these days ;)  I personally understand the reasons against it, but I've seen it in use in lots of places-

**Comment by pascalgrimaud** [comment]
@dhaalves-uel @cm325 : you can look at older issues like #398 #907 #1334 #2734 #3476

**Comment by deepu105** [comment]
@jdubois @cbornet sorry to beat around the dead horse but just wanted to ask. Since its bean a while that we did this and since no body complains should we make fluent setters the default with `public Bar setFoo()` syntax and remove the duplicate methods? 
I was looking at [this](https://github.com/jhipster/generator-jhipster/commit/2cd31621e76e872a5edf4d44405ec4bbb32ddd64) and thought it would have been nice to have fluent style everywhere

**Comment by jdubois** [comment]
I agree with you, but we don't know what people use: maybe most users still use setters... So, in doubt, I would keep them, even if I agree they should be removed

**Comment by deepu105** [comment]
@jdubois we are tracking it `insight.track('entity/fluentMethods', this.fluentMethods);` so may be we can check if anybody has turned it off and also for user class and other stuff we can start using fluent style.
For those who are using setters already they wouldn't see any difference if we start returning the object in that setter I guess

**Comment by jdubois** [comment]
I just checked, we have 0,01% of people not using fluent methods... If it doesn't do any harm (I haven't checked), let's remove this, OK.
But can we first do the Angular 4 code? This is less important

**Comment by cbornet** [comment]
Note : if we remove the setters then we don't respect the javabean contract anymore. Is it not an issue ? (Hibernate and Jackson don't rely on javabean. I think neither does mapstruct)

**Comment by cbornet** [comment]
What we can do is to have fluent setters by default that you can deactivate with --no-fluent-methods. And refactor the User classes.
@jdubois I can't really help on Angular (me back-end type 😅 ) so I can take this if you want.

**Comment by cbornet** [comment]
We also need to cover the UserMapper : our coverage decreased 😭

**Comment by jdubois** [comment]
@cbornet aren't they already fluent by default?

**Comment by cbornet** [comment]
@jdubois indeed !

**Comment by deepu105** [comment]
ofcourse I wanst suggesting we do it now. I just wanted to mention so that I dont forget. I mean we could create only fluent by default and the flag can be used to create non fluent if people want and for user classes we can have only fluent as they are jhipster domain classes. Anyways its not urgent at all.

**Pull request merged**